proposals/collection-literals.md

```diff
@@ -151,70 +156,76 @@ _Open issue: Relying on a parameter named `capacity` seems brittle. Is there an
+An error is reported for a collection literal if there are multiple `[CollectionBuilder]` attributes on the collection type.
 
-* For example, if an `init void Construct(T[] values)` instance method were added to [`ImmutableArray<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1), then it would be possible for the compiler to emit the following:
+The *create method* must be a `static` method defined on the *builder type*.
+The *create method* must have a single parameter of a type `System.ReadOnlySpan<T0>` passed by value, and a return value of the *collection type* to which the attribute is applied.
 
-    ```c#
-    T[] __storage = /* initialized using predefined rules */
-    ImmutableArray<T> __result = new ImmutableArray<T>();
-    __result.Construct(__storage);
-    ```
+The *builder type* cannot be generic.
+The *builder method* can be a generic *method* and the *arity* of the builder method must match the *arity* of the *collection type*.
 
-    `ImmutableArray<T>` would then take that array directly and use it as its own backing storage.  This would be safe because the compiler (following the requirements around `init`) would ensure that no other location in the code would have access to this temporary array, and thus it would not be possible to mutate it behind the back of the `ImmutableArray<T>` instance.
+Overloads with distinct signatures on the builder type, and overloads on base types, are ignored.
 
-    The above also demonstrates that this approach can work with struct types which do not have a [*parameterless struct constructor*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/parameterless-struct-constructors.md).  In the above, the call to `new ImmutableArray<T>()` is equivalent to `default(ImmutableArray<T>)`, (producing an `ImmutableArray<T>` whose [`IsDefault`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1.isdefault) property is initially true.  However, the `Construct` method can then safely update this to the final non-default state without that intermediate state being visible.
+An error is reported for a collection literal if the `[CollectionBuilder]` attribute on the collection type does not represent an invocable method with the expected signature defined on the *builder type*.
 
-* This formalization is quite beneficial because the only existing mechanism to (safely) create an ImmutableArray with values without copying is both excessively verbose and produces unavoidable garbage:
+If the parameter does not escape the *create method* (that is, if the parameter is `scoped` or if the return type is not a `ref struct`), the compiler *may* allocate the storage for the `ReadOnlySpan<T0>` on the stack rather than the heap.
```

Comment:
```suggestion
If the parameter does not escape the *create method* (that is, if the parameter is `scoped`, the return type is not a `ref struct` and there are no `out / ref` of type `ref struct`), the compiler will assign a lifetime of _current method_ to the argument. As such it *may* allocate the storage for the `ReadOnlySpan<T0>` on the stack rather than the heap. 
```

Reply:
Do not see any discussion about the lifetime of a collection literal that is used in a local declaration.

Reply:
I didn't mention `out` or `ref` parameters here because the *create method* pattern is specifically `static CollectionType<T> Create<T>(ReadOnlySpan<T> items)`, with a single parameter passed by value.

Reply:
True but shouldn't we discuss the more general policy here? Specifically that collection literals when passed to a method have such and such rules. The rules for custom create methods aren't special here, they just fall out from those broader rules.

Reply:
Basically there is no diff between 

```csharp

ImmutableArray<int> array1 = { 1, 2, 3 };
ImmutableArray<int> array2 = MyCreate({1, 2, 3 });
ImmutableArray<int> MyCreate(scoped ReadOnlySpan<int> span) => ...
```

Reply:
The comment here was to clarify the requirements for this particular case. We could perhaps remove this comment when we add a more general statement about lifetime in a subsequent PR.